<?xml version="1.0" ?>
<sdf version="1.10">
    <world name="swarm_world">
        <physics name="1ms" type="ignored">
            <max_step_size>0.001</max_step_size>
            <real_time_factor>1.0</real_time_factor>
        </physics>
        <plugin
            filename="gz-sim-physics-system"
            name="gz::sim::systems::Physics">
        </plugin>
        <plugin
            filename="gz-sim-user-commands-system"
            name="gz::sim::systems::UserCommands">
        </plugin>
        <plugin
            filename="gz-sim-scene-broadcaster-system"
            name="gz::sim::systems::SceneBroadcaster">
        </plugin>
        <plugin
            filename="gz-sim-sensors-system"
            name="gz::sim::systems::Sensors">
            <!-- TODO: figure out how to use ogre -->
            <render_engine>ogre</render_engine> 
        </plugin>

        <light type="directional" name="sun">
            <cast_shadows>true</cast_shadows>
            <pose>0 0 10 0 0 0</pose>
            <diffuse>0.8 0.8 0.8 1</diffuse>
            <specular>0.2 0.2 0.2 1</specular>
            <attenuation>
                <range>1000</range>
                <constant>0.9</constant>
                <linear>0.01</linear>
                <quadratic>0.001</quadratic>
            </attenuation>
            <direction>-0.5 0.1 -0.9</direction>
        </light>

        <gui fullscreen="1">
            <plugin filename="ImageDisplay" name="Image Display">
                <gz-gui>
                <title>Select Camera View</title>
                <property key="state" type="string">floating</property>
                <property type="double" key="width">800</property>
                <property type="double" key="height">800</property>
                </gz-gui>
                <topic>camera_world_view</topic>
                <topic_picker>true</topic_picker>
            </plugin>
            <plugin filename="WorldControl" name="World control">
                <gz-gui>
                <title>World control</title>
                <property type="bool" key="showTitleBar">false</property>
                <property type="bool" key="resizable">false</property>
                <property type="double" key="height">72</property>
                <property type="double" key="width">121</property>
                <property type="double" key="z">900</property>

                <property type="string" key="state">floating</property>
                <anchors target="3D View">
                    <line own="left" target="left"/>
                    <line own="bottom" target="bottom"/>
                </anchors>
                </gz-gui>

                <play_pause>true</play_pause>
                <step>true</step>
                <start_paused>true</start_paused>

            </plugin>
        </gui>

        <model name="camera">
            <pose>0 0 13 0 1.57 0</pose>
            <link name="link">
            <inertial>
                <mass>0</mass>
                <inertia>
                <ixx>0.000166667</ixx>
                <iyy>0.000166667</iyy>
                <izz>0.000166667</izz>
                </inertia>
            </inertial>
            <collision name="collision">
                <geometry>
                <box>
                    <size>0.1 0.1 0.1</size>
                </box>
                </geometry>
            </collision>
            <visual name="visual">
                <geometry>
                <box>
                    <size>0.1 0.1 0.1</size>
                </box>
                </geometry>
            </visual>
            <sensor name="camera" type="camera">
                <camera>
                <horizontal_fov>1.047</horizontal_fov>
                <image>
                    <width>1920</width>
                    <height>1080</height>
                </image>
                <clip>
                    <near>0.1</near>
                    <far>100</far>
                </clip>
                </camera>
                <always_on>1</always_on>
                <update_rate>60</update_rate>
                <visualize>true</visualize>
                <topic>camera_world_view</topic>
            </sensor>
            </link>
        </model>

        <model name="ground_plane">
            <static>true</static>
            <link name="link">
                <collision name="collision">
                <geometry>
                    <plane>
                    <normal>0 0 1</normal>
                    </plane>
                </geometry>
                </collision>
                <visual name="visual">
                <geometry>
                    <plane>
                    <normal>0 0 1</normal>
                    <size>1000 1000</size>
                    </plane>
                </geometry>
                <material>
                    <ambient>0.8 0.8 0.8 1</ambient>
                    <diffuse>0.8 0.8 0.8 1</diffuse>
                    <specular>0.8 0.8 0.8 1</specular>
                </material>
                </visual>
            </link>
        </model>

        <include>
            <uri>model://turtlefly</uri>      
            <name>turtle1</name>
            <pose>0 0 0 0 0 0</pose>
        </include>

        <include>
            <uri>model://turtlefly</uri>      
            <name>turtle2</name>
            <pose>1 0 0 0 0 -1.5</pose>
        </include>

        <include>
            <uri>model://turtlefly</uri>      
            <name>turtle3</name>
            <pose>2 0 0 0 0 2.5</pose>
        </include>


        <include>
            <uri>model://turtlefly</uri>      
            <name>turtle4</name>
            <pose>0 1 0 0 0 1.5</pose>
        </include>

        <include>
            <uri>model://turtlefly</uri>      
            <name>turtle5</name>
            <pose>0 2 0 0 0 -3</pose>
        </include>
        
        <include>
            <uri>model://turtlefly</uri>      
            <name>turtle6</name>
            <pose>1 1 0 0 0 -2</pose>
        </include>

        <include>
            <uri>model://turtlefly</uri>      
            <name>turtle7</name>
            <pose>1 2 0 0 0 -1</pose>
        </include>

        <include>
            <uri>model://turtlefly</uri>      
            <name>turtle8</name>
            <pose>2 1 0 0 0 3</pose>
        </include>


        <include>
            <uri>model://turtlefly</uri>      
            <name>turtle9</name>
            <pose>2 2 0 0 0 2</pose>
        </include>

        <include>
            <uri>model://turtlefly</uri>      
            <name>turtle10</name>
            <pose>0.5 0.5 0 0 0 1</pose>
        </include>

    <model name='wall1'>
      <pose>3 0 0 0 0 0</pose>
      <link name='box_link'>
        <inertial>
          <inertia>
            <ixx>0.16666</ixx>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyy>0.16666</iyy>
            <iyz>0</iyz>
            <izz>0.16666</izz>
          </inertia>
          <mass>1</mass>
          <pose>0 0 0 0 0 0</pose>
        </inertial>
        <collision name='box_collision'>
          <geometry>
            <box>
              <size>1 10 1</size>
            </box>
          </geometry>
          <surface>
            <friction>
              <ode/>
            </friction>
            <bounce/>
            <contact/>
          </surface>
        </collision>
        <visual name='box_visual'>
          <geometry>
            <box>
              <size>1 16 1</size>
            </box>
          </geometry>
          <material>
            <ambient>0.300000012 0.300000012 0.300000012 1</ambient>
            <diffuse>0.699999988 0.699999988 0.699999988 1</diffuse>
            <specular>1 1 1 1</specular>
          </material>
        </visual>
        <pose>0 0 0 0 0 0</pose>
        <enable_wind>false</enable_wind>
      </link>
      <static>true</static>
      <self_collide>false</self_collide>
    </model> 

    <model name='wall2'>
      <pose>-3 0 0 0 0 0</pose>
      <link name='box_link'>
        <inertial>
          <inertia>
            <ixx>0.16666</ixx>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyy>0.16666</iyy>
            <iyz>0</iyz>
            <izz>0.16666</izz>
          </inertia>
          <mass>1</mass>
          <pose>0 0 0 0 0 0</pose>
        </inertial>
        <collision name='box_collision'>
          <geometry>
            <box>
              <size>1 10 1</size>
            </box>
          </geometry>
          <surface>
            <friction>
              <ode/>
            </friction>
            <bounce/>
            <contact/>
          </surface>
        </collision>
        <visual name='box_visual'>
          <geometry>
            <box>
              <size>1 16 1</size>
            </box>
          </geometry>
          <material>
            <ambient>0.300000012 0.300000012 0.300000012 1</ambient>
            <diffuse>0.699999988 0.699999988 0.699999988 1</diffuse>
            <specular>1 1 1 1</specular>
          </material>
        </visual>
        <pose>0 0 0 0 0 0</pose>
        <enable_wind>false</enable_wind>
      </link>
      <static>true</static>
      <self_collide>false</self_collide>
    </model> 

    <model name='wall3'>
      <pose>0 -4 0 0 0 0</pose>
      <link name='box_link'>
        <inertial>
          <inertia>
            <ixx>0.16666</ixx>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyy>0.16666</iyy>
            <iyz>0</iyz>
            <izz>0.16666</izz>
          </inertia>
          <mass>1</mass>
          <pose>0 0 0 0 0 0</pose>
        </inertial>
        <collision name='box_collision'>
          <geometry>
            <box>
              <size>8 1 1</size>
            </box>
          </geometry>
          <surface>
            <friction>
              <ode/>
            </friction>
            <bounce/>
            <contact/>
          </surface>
        </collision>
        <visual name='box_visual'>
          <geometry>
            <box>
              <size>8 1 1</size>
            </box>
          </geometry>
          <material>
            <ambient>0.300000012 0.300000012 0.300000012 1</ambient>
            <diffuse>0.699999988 0.699999988 0.699999988 1</diffuse>
            <specular>1 1 1 1</specular>
          </material>
        </visual>
        <pose>0 0 0 0 0 0</pose>
        <enable_wind>false</enable_wind>
      </link>
      <static>true</static>
      <self_collide>false</self_collide>
    </model> 

    <model name='wall4'>
      <pose>0 4 0 0 0 0</pose>
      <link name='box_link'>
        <inertial>
          <inertia>
            <ixx>0.16666</ixx>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyy>0.16666</iyy>
            <iyz>0</iyz>
            <izz>0.16666</izz>
          </inertia>
          <mass>1</mass>
          <pose>0 0 0 0 0 0</pose>
        </inertial>
        <collision name='box_collision'>
          <geometry>
            <box>
              <size>8 1 1</size>
            </box>
          </geometry>
          <surface>
            <friction>
              <ode/>
            </friction>
            <bounce/>
            <contact/>
          </surface>
        </collision>
        <visual name='box_visual'>
          <geometry>
            <box>
              <size>8 1 1</size>
            </box>
          </geometry>
          <material>
            <ambient>0.300000012 0.300000012 0.300000012 1</ambient>
            <diffuse>0.699999988 0.699999988 0.699999988 1</diffuse>
            <specular>1 1 1 1</specular>
          </material>
        </visual>
        <pose>0 0 0 0 0 0</pose>
        <enable_wind>false</enable_wind>
      </link>
      <static>true</static>
      <self_collide>false</self_collide>
    </model> 
        
    </world>
</sdf>